test/test_torch.py

```diff
@@ -5376,6 +5376,283 @@ def test_narrow_empty(self, device):
+            constant_values = torch.zeros(constant_values_size, device=device, dtype=constant_values_dtype)
+
+            result_torch = input.pad(pad_widths, constant_values=constant_values).cpu()
+
+            result_torch_precast = input.pad(pad_widths, constant_values=constant_values.to(dtype)).cpu()
+            self.assertEqual(result_torch, result_torch_precast)
+
+            if torch.bfloat16 not in [dtype, constant_values_dtype]:
+                # numpy.pad has weird inconsistent support for dtype mismatches
+                # between the input and constant_values. Mismatches are
+                # supported in most cases, but when both of the following are
+                # conditions are true, numpy.pad throws an error
+                size_unsupported_condition = constant_values.size() in [torch.Size([input.dim(), 1]), torch.Size([input.dim(), 2])]
+                dtypes_unsupported_condition = constant_values_dtype.is_complex and not dtype.is_complex and (dtype != torch.bool)
+
+                if not size_unsupported_condition or not dtypes_unsupported_condition:
+                    result_numpy = torch.tensor(np.pad(
+                        input.cpu().numpy(),
+                        pad_widths,
+                        constant_values=constant_values.cpu().numpy()))
+                    self.assertEqual(result_torch, result_numpy)
+
+    @onlyOnCPUAndCUDA
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_pad_errors(self, device, dtype):
+        input = make_tensor((3, 3, 3), device, dtype, low=-9, high=9)
+
+        # pad_width tensor dtype errors
+        for pad_width_dtype in torch.testing.get_all_dtypes():
+            for pad_width_size in [(), (2,), (input.dim(), 2)]:
+                pad_width = torch.zeros(pad_width_size, device='cpu', dtype=pad_width_dtype)
+
+                if pad_width_dtype == torch.int64:
+                    input.pad(pad_width)
+                    if torch.bfloat16 not in [dtype, pad_width_dtype]:
+                        np.pad(input.cpu().numpy(), pad_width.cpu().numpy())
+                else:
+                    with self.assertRaisesRegex(RuntimeError, r"Expected 'pad_width' to be Long dtype"):
+                        input.pad(pad_width)
+
+                    if torch.bfloat16 not in [dtype, pad_width_dtype]:
+                        # Numpy does support char, short, and int types, but torch.pad only supports long
+                        if pad_width_dtype not in [torch.int8, torch.int16, torch.int32]:
+                            with self.assertRaises(TypeError):
+                                np.pad(input.cpu().numpy(), pad_width.cpu().numpy())
+
+        # pad_width primitive type errors
+        for pad_width in [0., (0.,), ((0, 0), (0, 0), (0., 0)), 1j, (1j,), ((0, 0), (0, 0), (1j, 0))]:
+            with self.assertRaisesRegex(RuntimeError, r"Expected 'pad_width' to be of integer type"):
```

Comment:
This section will probably need to be extended with testing that a constant pad value isn't 1.5 or inf or nan when padding an integer tensor